Fix #1024 : Exploration back flow

Sarthak2601 opened the pull request:
<!--
  - Thanks for submitting code to Oppia! Please fill out the following as part of
  - your pull request so we can review your code more easily.
  -->

## Explanation
<!--
  - Explain what your PR does. If this PR fixes an existing bug, please include
  - "Fixes #bugnum:" in the explanation so that GitHub can auto-close the issue
  - when this PR is merged.
  -->
- Fix #1024 

- The TopicFragmentPresenter only set the Lession tab as the starting tab of the viewPager if both topicID and storyID returned .isNotEmpty() as true. But since story id wasn't getting returned by the stopExploration() so it set the default tab (Info) for viewPager.

- I also added the onBackPressed() to the Topic Activity so that it doesn't create any further bugs like unnecessary re-opening of topics(as opened by user) on back pressed. 

- I would like to know if I should also implement an AlertDialog on the back press of the Topic Activity ? 

## Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is made from a branch that is up-to-date with "develop".
- [x] The PR's branch is based on "develop" and not on any other branch.
- [x] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

Sarthak2601 committed:
exploration back press

rt4914 reviewed (changes requested):
Following flows are incorrect:

1. Click on Topic on homescreen should open `Info` page and not `PromotedStory`
![Screenshot_1588566559](https://user-images.githubusercontent.com/9396084/80936280-160e7300-8dee-11ea-82d5-a5a4f1e6e3ed.png)


2. Open Lessons tab -> Click on an item to open `StoryActivity` which contains the chapter list with images. Click on chapter - exploration starts. Leave exploration. In this case `StoryActivity` should appear instead of `TopicActivity` with Lessons tab.
![story_activity](https://user-images.githubusercontent.com/9396084/80936364-80271800-8dee-11ea-801c-c9563c6eb9d0.gif)


All other flows are correct but these two flows are still incorrect. Maybe you can introduce two different enum to decide these screens:
1. For deciding which activity to open in back flow
```
enum class ParentActivityForExploration(val value: Int) {
  STORY_ACTIVITY(0),
  TOPIC_ACTIVITY(1);
}

```

2. For deciding which tab to open in TopicActivity in back flow
```
enum class TopicActivityStartingTab(val value: Int) {
  INFO(0),
  LESSONS(1);
}
```

You can pass these enums to `ExplorationActivity` and use this information to control which screen to show on back press.

Sarthak2601 commented:
> Following flows are incorrect:
> 
> 1. Click on Topic on homescreen should open `Info` page and not `PromotedStory`
>    ![Screenshot_1588566559](https://user-images.githubusercontent.com/9396084/80936280-160e7300-8dee-11ea-82d5-a5a4f1e6e3ed.png)
> 2. Open Lessons tab -> Click on an item to open `StoryActivity` which contains the chapter list with images. Click on chapter - exploration starts. Leave exploration. In this case `StoryActivity` should appear instead of `TopicActivity` with Lessons tab.
>    ![story_activity](https://user-images.githubusercontent.com/9396084/80936364-80271800-8dee-11ea-801c-c9563c6eb9d0.gif)
> 
> All other flows are correct but these two flows are still incorrect. Maybe you can introduce two different enum to decide these screens:
> 
> 1. For deciding which activity to open in back flow
> 
> ```
> enum class ParentActivityForExploration(val value: Int) {
>   STORY_ACTIVITY(0),
>   TOPIC_ACTIVITY(1);
> }
> ```
> 
> 1. For deciding which tab to open in TopicActivity in back flow
> 
> ```
> enum class TopicActivityStartingTab(val value: Int) {
>   INFO(0),
>   LESSONS(1);
> }
> ```
> 
> You can pass these enums to `ExplorationActivity` and use this information to control which screen to show on back press.

Sorry I took a pass there, will wrap it up soon !

Sarthak2601 committed:
back press flow correction 01

Sarthak2601 committed:
back press flow correction 02

Sarthak2601 committed:
back press flow correction 03

Sarthak2601 commented:
@rt4914 Both the flows are now corrected. I also added onBackPressed() to the StoryActivity as there was some error in that as well. 
PTAL, thanks :)

rt4914 commented:
> @rt4914 Both the flows are now corrected. I also added onBackPressed() to the StoryActivity as there was some error in that as well.
> PTAL, thanks :)

@Sarthak2601
Application is crashing for this flow:
`Home -> View All -> RecentlyPlayedActivity -> Select any item` It will crash the application as the "exploration_id" is null.

rt4914 commented:
> > @rt4914 Both the flows are now corrected. I also added onBackPressed() to the StoryActivity as there was some error in that as well.
> > PTAL, thanks :)
> 
> Application is crashing for this flow:
> `Home -> View All -> RecentlyPlayedActivity -> Select any item` It will crash the application as the "exploration_id" is null.

Sarthak2601 committed:
flow correction

Sarthak2601 committed:
flow correction

Sarthak2601 committed:
Merge branch 'develop' into exploration-back-flow

Sarthak2601 commented:
@rt4914 Handled the crash and corrected the back flow of the Exploration activity opened via RecentlyPlayed stories to open the info tab as well.

rt4914 commented on app/src/main/java/org/oppia/app/home/recentlyplayed/RecentlyPlayedFragmentPresenter.kt:
```diff
@@ -98,7 +98,7 @@ class RecentlyPlayedFragmentPresenter @Inject constructor(
         else -> {
           logger.d("RecentlyPlayedFragment", "Successfully loaded exploration")
-          routeToExplorationListener.routeToExploration(internalProfileId, topicId, storyId, explorationId)
+          routeToExplorationListener.routeToExploration(internalProfileId, topicId, storyId, explorationId,null)
```
Nit: whenever we pass null or an integer or a boolean value we write inline comment to make it more readable code.
`routeToExplorationListener.routeToExploration(internalProfileId, topicId, storyId, explorationId, /* backflowId= */ null)`

Sarthak2601 replied:
Okay, done :)

rt4914 commented on app/src/main/java/org/oppia/app/player/exploration/ExplorationActivity.kt:
```diff
@@ -38,7 +39,8 @@ class ExplorationActivity : InjectableAppCompatActivity(), StopExplorationInterf
     explorationId = intent.getStringExtra(EXPLORATION_ACTIVITY_EXPLORATION_ID_ARGUMENT_KEY)
-    explorationActivityPresenter.handleOnCreate(internalProfileId, topicId, storyId, explorationId)
+    backflowId = intent.getStringExtra(EXPLORATION_ACTIVITY_BACKFLOW_ID)
+    explorationActivityPresenter.handleOnCreate(internalProfileId, topicId, storyId, explorationId, backflowId, this)
```
`Context` should be passed as first parameter instead of last.

Sarthak2601 replied:
Done :)

rt4914 reviewed (changes requested):
Great implementation. Your entire code approach is correct but instead of using string constants you should use `enum` for backflow ids.

You can check `AdminAuthEnum` and how it is being used in similar way.

Also, instead of `backflowId` you should name it as `backflowScreen` because the value is a screen name and not an id.

Sarthak2601 committed:
enum implementation + nits

Sarthak2601 commented:
> Great implementation. Your entire code approach is correct but instead of using string constants you should use `enum` for backflow ids.
> 
> You can check `AdminAuthEnum` and how it is being used in similar way.
> 
> Also, instead of `backflowId` you should name it as `backflowScreen` because the value is a screen name and not an id.

@rt4914 Done with all the changes, PTAL :)

rt4914 commented on app/src/main/java/org/oppia/app/home/recentlyplayed/RecentlyPlayedFragmentPresenter.kt:
```diff
@@ -98,7 +98,7 @@ class RecentlyPlayedFragmentPresenter @Inject constructor(
         else -> {
           logger.d("RecentlyPlayedFragment", "Successfully loaded exploration")
-          routeToExplorationListener.routeToExploration(internalProfileId, topicId, storyId, explorationId)
+          routeToExplorationListener.routeToExploration(internalProfileId, topicId, storyId, explorationId, /* backflowScreen = */null)
```
Nit: Spacing issue there should be space before `null`
`/* backflowScreen = */ null`

Sarthak2601 replied:
Done

rt4914 commented on app/src/main/java/org/oppia/app/player/exploration/ExplorationActivity.kt:
```diff
@@ -38,7 +39,8 @@ class ExplorationActivity : InjectableAppCompatActivity(), StopExplorationInterf
     storyId = intent.getStringExtra(EXPLORATION_ACTIVITY_STORY_ID_ARGUMENT_KEY)
     explorationId = intent.getStringExtra(EXPLORATION_ACTIVITY_EXPLORATION_ID_ARGUMENT_KEY)
-    explorationActivityPresenter.handleOnCreate(internalProfileId, topicId, storyId, explorationId)
+    backflowScreen = intent.getIntExtra(EXPLORATION_ACTIVITY_BACKFLOW_SCREEN,-1)
```
Spacing issue: space after `,`
`intent.getIntExtra(EXPLORATION_ACTIVITY_BACKFLOW_SCREEN ,-1)`

Sarthak2601 replied:
Done

rt4914 commented on app/src/main/java/org/oppia/app/player/exploration/ExplorationActivity.kt:
```diff
@@ -48,18 +50,22 @@ class ExplorationActivity : InjectableAppCompatActivity(), StopExplorationInterf
     internal const val EXPLORATION_ACTIVITY_TOPIC_ID_ARGUMENT_KEY = "ExplorationActivity.topic_id"
     internal const val EXPLORATION_ACTIVITY_STORY_ID_ARGUMENT_KEY = "ExplorationActivity.story_id"
     internal const val EXPLORATION_ACTIVITY_EXPLORATION_ID_ARGUMENT_KEY = "ExplorationActivity.exploration_id"
+    internal const val EXPLORATION_ACTIVITY_BACKFLOW_SCREEN = "ExplorationActivity.backflow_screen"
```
Rename this to `EXPLORATION_ACTIVITY_BACKFLOW_SCREEN_KEY`

Sarthak2601 replied:
Done

rt4914 commented on app/src/main/java/org/oppia/app/player/exploration/ExplorationActivityPresenter.kt:
```diff
@@ -131,6 +145,14 @@ class ExplorationActivityPresenter @Inject constructor(
 
+  private fun backPressActivitySelector(backflowScreen: Int?){
+    when(backflowScreen){
+      ParentActivityForExploration.BACKFLOW_SCREEN_STORY.value -> activity.startActivity(StoryActivity.createStoryActivityIntent(context,internalProfileId,topicId,storyId))
```
Spacing issue: `createStoryActivityIntent(context, internalProfileId, topicId, storyId)`

rt4914 replied:
Ditto for below two lines, there should be space after `,`

Sarthak2601 replied:
Done

rt4914 commented on app/src/main/java/org/oppia/app/story/StoryFragment.kt:
```diff
@@ -48,8 +48,8 @@ class StoryFragment : InjectableFragment(), ExplorationSelectionListener {
-  override fun selectExploration(internalProfileId: Int, topicId: String, storyId: String, explorationId: String) {
-    storyFragmentPresenter.handleSelectExploration(internalProfileId, topicId, storyId, explorationId)
+  override fun selectExploration(internalProfileId: Int, topicId: String, storyId: String, explorationId: String, backflowId: Int?) {
+    storyFragmentPresenter.handleSelectExploration(internalProfileId, topicId, storyId, explorationId, backflowId)
```
Rename `backflowId ` to `backflowScreen` here and everywhere else.

Sarthak2601 replied:
Done

rt4914 commented on app/src/main/java/org/oppia/app/story/StoryFragmentPresenter.kt:
```diff
@@ -68,8 +68,8 @@ class StoryFragmentPresenter @Inject constructor(
-  fun handleSelectExploration(internalProfileId: Int, topicId: String, storyId: String, explorationId: String) {
-    routeToExplorationListener.routeToExploration(internalProfileId, topicId, storyId, explorationId)
+  fun handleSelectExploration(internalProfileId: Int, topicId: String, storyId: String, explorationId: String, backflowId: Int?) {
+    routeToExplorationListener.routeToExploration(internalProfileId, topicId, storyId, explorationId,backflowId)
```
Rename `backflowId ` to `backflowScreen` here and everywhere else.

rt4914 replied:
Also check the space issue before `backflowId `

Sarthak2601 replied:
Done

rt4914 commented on app/src/main/java/org/oppia/app/story/storyitemviewmodel/StoryChapterSummaryViewModel.kt:
```diff
@@ -39,7 +40,7 @@ class StoryChapterSummaryViewModel(
         else -> {
           logger.d(STORY_VIEWER_TAG, "Successfully loaded exploration: $explorationId")
-          explorationSelectionListener.selectExploration(internalProfileId, topicId, storyId, explorationId)
+          explorationSelectionListener.selectExploration(internalProfileId, topicId, storyId, explorationId, /* backflowScreen = */1 )
```
Spacing issue:
`explorationSelectionListener.selectExploration(internalProfileId, topicId, storyId, explorationId, /* backflowScreen = */ 1)`

Sarthak2601 replied:
Done

rt4914 commented on app/src/main/java/org/oppia/app/testing/ExplorationTestActivity.kt:
```diff
@@ -17,14 +17,15 @@ class ExplorationTestActivity : InjectableAppCompatActivity(), RouteToExploratio
   }
 
-  override fun routeToExploration(internalProfileId: Int, topicId: String, storyId: String, explorationId: String) {
+  override fun routeToExploration(internalProfileId: Int, topicId: String, storyId: String, explorationId: String, backflowScreen: Int? ) {
```
Spacing issue: Remove space before `)`
`override fun routeToExploration(internalProfileId: Int, topicId: String, storyId: String, explorationId: String, backflowScreen: Int?)`

Sarthak2601 replied:
Done

rt4914 commented on app/src/main/java/org/oppia/app/testing/ExplorationTestActivityPresenter.kt:
```diff
@@ -48,7 +48,7 @@ class ExplorationTestActivityPresenter @Inject constructor(
         else -> {
           logger.d(TAG_EXPLORATION_TEST_ACTIVITY, "Successfully loaded exploration")
-          routeToExplorationListener.routeToExploration(INTERNAL_PROFILE_ID, TOPIC_ID, STORY_ID, EXPLORATION_ID)
+          routeToExplorationListener.routeToExploration(INTERNAL_PROFILE_ID, TOPIC_ID, STORY_ID, EXPLORATION_ID,/* backflowScreen= */null)
```
Spacing issue: `routeToExplorationListener.routeToExploration(INTERNAL_PROFILE_ID, TOPIC_ID, STORY_ID, EXPLORATION_ID, /* backflowScreen= */ null)`

Sarthak2601 replied:
Done

rt4914 commented on app/src/main/java/org/oppia/app/topic/TopicActivity.kt:
```diff
@@ -48,18 +49,23 @@ class TopicActivity : InjectableAppCompatActivity(), RouteToQuestionPlayerListen
   }
 
+  override fun onBackPressed() {
+    startActivity(HomeActivity.createHomeActivity(this,internalProfileId))
```
Spacing issue: `startActivity(HomeActivity.createHomeActivity(this, internalProfileId))`

Sarthak2601 replied:
Done

rt4914 commented on app/src/main/java/org/oppia/app/topic/lessons/TopicLessonsFragmentPresenter.kt:
```diff
@@ -118,10 +119,10 @@ class TopicLessonsFragmentPresenter @Inject constructor(
 
   override fun selectChapterSummary(storyId: String, chapterSummary: ChapterSummary) {
-    playExploration(internalProfileId, topicId, storyId, chapterSummary.explorationId)
+    playExploration(internalProfileId, topicId, storyId, chapterSummary.explorationId, /*backflowScreen =*/0)
```
Spacing issue : `/*backflowScreen =*/ 0`

Sarthak2601 replied:
Done

rt4914 commented on app/src/sharedTest/java/org/oppia/app/player/exploration/ExplorationActivityTest.kt:
```diff
@@ -285,7 +285,7 @@ class ExplorationActivityTest {
   private fun createExplorationActivityIntent(internalProfileId: Int, topicId: String, storyId: String, explorationId: String): Intent {
     return ExplorationActivity.createExplorationActivityIntent(
-      ApplicationProvider.getApplicationContext(), internalProfileId, topicId, storyId, explorationId
+      ApplicationProvider.getApplicationContext(), internalProfileId, topicId, storyId, explorationId,null
```
Spacing and comment issue: `explorationId, /* backflowScreen= */ null`

Sarthak2601 replied:
Done

rt4914 reviewed (changes requested):
All changes are based on spacing issue, naming issue only. Update these and also self-review this PR once this is done, so that it becomes easy to merge.

Sarthak2601 committed:
exploration back flow : nit

Sarthak2601 committed:
exploration back flow : nits

Sarthak2601 commented:
@rt4914 Done with all the nits. Self reviewed and removed unused imports and did one more nit as well. PTAL :)

rt4914 reviewed (approved):
@Sarthak2601 LGTM

rt4914 commented:
@Sarthak2601 Just resolve merge conflict and merge it. Thanks.

Sarthak2601 committed:
Merge branch 'develop' into exploration-back-flow
